autoload/maktaba/syscall.vim

```diff
@@ -89,6 +103,59 @@ function! maktaba#syscall#DoCall() abort dict
   return l:return_data
 endfunction
 
+
+""
+" @public
+" Returns whether the current vim session supports asynchronous calls.
+function! maktaba#syscall#IsAsyncAvailable()
+  return !s:async_disabled && !empty(v:servername) && has('clientserver')
+endfunction
+
+
+""
+" @private
+" @dict Syscall
+" Calls |system()| asynchronously, and invokes a @function(this.callback) once
+" the command completes, passing in stdout, stderr and exit code to it.
+" The specific implementation for @function(#CallAsync).
+function! maktaba#syscall#DoCallAsync() abort dict
+  if !maktaba#syscall#IsAsyncAvailable()
+    if self.allow_sync_fallback
+      let l:return_data = self.Call()
```

Comment:
Since this branch is likely to surprise users with delays, let's at least log a maktaba warning, something like
```vim
call s:plugin.logger.Warn('Async support not available. ' .
    \ 'Falling back to synchronous execution for system call: ' .
    \ self.GetCommand())
```
(with `s:plugin` set to `maktaba#Maktaba()` at the top of the file).